### ISSUE TITLE
Traffic shaping

### ISSUE THREAD

**Issue by lhstrh** [icyphy/lingua-franca#98]
TL;DR: 
1. Traffic shaping is at odds with the `additionalDelay` parameter of `schedule`; should we ignore it for physical actions?
2. Could we get rid of the delay for physical actions altogether? I don't see it used anywhere.

We now enforce a minimum inter-arrival time for physical actions that specify one.

For an action that is defined as `physical action a(0, 10 msec)`, when it first gets scheduled at physical time `T`, the tag will be `T`. If it then subsequently gets scheduled at `T+1 msec`, the tag of the event will be `T+ 10 msec`; if then it gets scheduled at `T+2 msec`, the tag of the event will be `T+ 20 msec`, an so forth. Only once there are no events for `a` in the queue that have a tag that is within `10 msec` of T, the assigned timestamp will again be `T`. 

The issue is that physical actions are not necessarily scheduled with monotonically increasing tags, because `schedule` accepts an `additionalDelay` parameter. Satisfying a minimum inter-arrival time in such non-monotonic scenarios is non-trivial. It will require expensive searches in the event queue to determine whether inserting an event with tag T will violate the constraint. And if it does, what should be done if adding a delay to the tag will put it too close to another event that is already in the queue?

The `additionalDelay` was originally introduced to express "variable delays," which is useful for logical actions. But is it useful for physical actions? I'm not so sure... Maybe we should ignore the `additionalDelay` parameter for physical actions? Going even further, do we need a `minDelay` for physical actions at all? If we don't, that can save us a parameter in the definition of actions. Because traffic shaping is not done for logical actions, we could have both types of actions have a single parameter but with a different meaning. For logical actions, the parameter would denote a minimum delay (as it currently does), while for physical actions the parameter would denote a minimum inter-arrival time.

**Comment by edwardalee** [comment]
Interesting observation. Yes, I think your solution is right, except that instead of ignoring the additionalDelay argument when scheduling a physical action, couldn't this additionalDelay be interpreted as adding temporarily to the minimum inter arrival time?  This would keep monotonicity, and ignoring arguments altogether makes me nervous.

**Comment by edwardalee** [comment]
Also, the grammar now allows physical and logical actions to have two parameters, minDelay and minInterArrival. I suggest we reduce this to just one. For logical actions, it is a minDelay. For physical actions it is a minInterArrival.  In the grammar, we could rename it minTime.  Then the usage will be consistent with the argument to schedule, which will be interpreted in these two same ways.

**Comment by lhstrh** [comment]
I like the idea of treating the argument to schedule as a temporary increase of the minimum inter arrival time.

We're also on the same page about having a single parameter for logical and physical actions alike. We have to be certain, however, that we have no use for traffic shaping of events for logical actions. It could be done, of course; the only difference with physical actions would be that the default minimum inter arrival would be 0. It must be understood that with the syntax proposal we have on the table right now we would eliminate that feature.

I think it would still be useful to indicate a particular traffic shaping policy; while, by default, we could delay, it would be convenient if we could specify alternative policies. In particular, it seems useful to have a "replace" (i.e., of the events that succeed too rapidly only the last one is kept) or "drop" (i.e., of the events that succeed too rapidly only the first one is kept). The advantage of these policies is that they limit the stress on event queue. Perhaps we could indicate the policy as a second, optional parameter of physical actions?

**Comment by edwardalee** [comment]
I don’t think we will lose anything by reducing physical actions to one parameter.  An additional delay could always be accomplished by chaining a logical action to occur after a physical one.  But I’ve had a hard time coming up with any scenario where one would want to do that.  So I don’t think it’s worth complicating the syntax for such an esoteric usage pattern.

**Comment by edwardalee** [comment]
One subtlety is that how to handle the minimum interarrival time D for the *first* appearance of a physical action. I think what we should do is delay the first appearance to occur no earlier than D after the start of execution. This makes it more similar to the minDelay argument of the logical action during startup, and I think it also makes sense for applications: startup is a difficult time, often with lots of work to do once, so delaying sporadic stimulus makes sense to me.

**Comment by edwardalee** [comment]
OK, I have implemented this. More testing is needed.

**Comment by edwardalee** [comment]
I have added tests.

**Comment by lhstrh** [comment]
This looks good! I'm just reopening the ticket because the second part of my question remains unaddressed:

> I think it would still be useful to indicate a particular traffic shaping policy; while, by default, we could delay, it would be convenient if we could specify alternative policies. In particular, it seems useful to have a "replace" (i.e., of the events that succeed too rapidly only the last one is kept) or "drop" (i.e., of the events that succeed too rapidly only the first one is kept). The advantage of these policies is that they limit the stress on event queue. Perhaps we could indicate the policy as a second, optional parameter of physical actions?

We could have something like: `physical action(3 ms, DROP) a;` or  `physical action(3 ms, REPLACE) b;` and the default would be `physical action(3 ms, DELAY) c;`

**Comment by lhstrh** [comment]
See [23c5a5](23c5a5).

**Comment by edwardalee** [comment]
What exactly would REPLACE mean?  What if the previous event is no longer on the event queue?

**Comment by lhstrh** [comment]
Good question! I haven't implemented the behavior yet, so I'm not sure... But when the previous event is no longer on the queue, then I think the only sensible thing left to do is delay the new event. This approximates having the "most recent value" available at all times. Perhaps a better term would be `UPDATE`?

**Comment by lhstrh** [comment]
The first sentence of the original post is still an issue: 
> Traffic shaping is at odds with the additionalDelay parameter of schedule

With the capability of scheduling events in a non-monotonic fashion (i.e., the tags of scheduled events are not monotonically increasing), a newly-scheduled event can collide (i.e., be closer to another event than is permitted by the MIT) with events that can be either ahead or behind it on the logical time line. Under a `DROP` policy, this would not necessarily be problematic. However, under a `DEFER` policy, which adjusts the tag so as to not violate the MIT, it may require traversing the entire event queue before finding out that the next available "slot" is a the very end of the queue. This is problematic from a performance, as well as a semantic perspective.

There are different ways of addressing this problem.
- We could remove the `DEFER` policy and only implement `DROP`.
- We could ignore the `additionalDelay` parameter in `schedule` if the involved action has an MIT specified. This would guarantee monotonic behavior and make the `DEFER` policy simple and efficient. Ignored parameters, however, are not a staple of a clean an intuitive API.
- We could remove the `additionalDelay` parameter from `schedule` altogether. This would not be backward compatible, but would simplify the API. To achieve variable delays, we could augment actions with a setter to modify intrinsic delay parameter. We would have to distinguish these variable delays from constant ones, and only for those ones would it be possible to specify an MIT. For example, we could have a `variable-delay` modifier in LF, to appear in front of an action to declare its delay to be variable (and forbid the specification of an MIT).
  - This solution gives rise to another question though: should it be possible to set the action's delay parameter to anything smaller than what it originally was set to? If we allow this, we'd have to assume a minimum delay of zero for variable-delay actions in any static analysis that we do.
  - Note that we could also make constant delays the exception and introduce a `constant-delay` modifier (and disallow the specification of an MIT if the modifier is missing). This might lead to verbosity in LF code, because in most uses the delays are constants.
- We could change the meaning of MIT to refer to the physical time interval between successive calls to schedule with a particular action. This, however, in combination with an `additionalDelay` parameter would still allow for events to be scheduled with tags that are very close to one another. This would render the mechanism useless as a mechanism for preventing system overload.

**Comment by edwardalee** [comment]
I think you overstate the downsides in the last bullet.  The long-term average occurrence rate of the action will still be determined by the MIT.  Also, it's not "useless" because the common case will be to have a zero delay.  Moreover, "system overload" is a physical-time problem, not a logical-time problem, so it actually makes sense to me that a property that is all about preventing system overload should reference the physical timeline.

**Comment by lhstrh** [comment]
If the runtime does not advance logical time until physical time exceeds the tag of the next event, then the release time of the events, albeit placed on a logical time line, coincide with the physical timeline, so I'm not sure it really make sense to make that distinction here. One could have three rapidly succeeding calls to schedule that scatter events over a large interval of logical time, or place them arbitrarily close to one another. Therefore, it doesn't make sense to me to impose a restriction on the time in between successive calls (call it a minimum interscheduling time); this approach may be too restrictive in some cases, too relaxed in others.

**Comment by edwardalee** [comment]
This really depends on deadlines.  If there are no such constraints, then average rate is sufficient to ensure correctness (given enough memory).

**Comment by lhstrh** [comment]
I see, so the issue is a bit more nuanced. Nonetheless, there are quite a number of options in the list to consider... Seems like a good discussion topic for tomorrow's telecon.

**Comment by lhstrh** [comment]
This issue has been lingering for a long time, and I think we finally have a good candidate solution. Please provide feedback if you have any. We have to carefully evaluate this proposal. While it does align with the goals of the language, the changes are not backward compatible. The proposed changes can be summarized as follows:

- Actions (logical or physical) are no longer parameterized with a delay, but with a minimum interarrival time (MIT);
- The parameter given to `schedule` will be renamed from `additionalDelay` to simply `delay`. It will be the sole mechanism to specify a logical time delay for a scheduled event;
- MIT is defined as the distance between tags of subsequently scheduled events. It has to be greater than or equal to zero. The default is zero. This definition guarantees monotonicity and implies that a violation of a specified MIT is always due to an event with an _earlier_ tag. This makes the implementation of error handling policies trivial and inexpensive. An MIT is considered violated when the computed tag of a newly scheduled event `t_new` on some action `a` is closer to the tag of the last scheduled event `t_last` on `a` than is allowed by the MIT specified by `a`; and
- We define the following policies (to be specified as a second and optional parameter of an action) for handling MIT violations: `DROP` (keep the conflicting event, do not enqueue a new event) , `DEFER` (enqueue the new event, but adjust its tag so that `t_new = t_last + MIT`, and `REPLACE` or `UPDATE` (replace the value of the conflicting event)

It remains that case that if the MIT is zero and an event is scheduled on the same action with the same tag, the value is overwritten (just like we do with ports).

Some open questions remain:
- What should be the default policy? Perhaps `DEFER` since it preserves all data.
- What should `schedule` return if it encountered an MIT violation? Perhaps the tag of the event in which it recorded the specified value? Or maybe the earliest tag at which a subsequent event can be scheduled without causing an MIT violation? The latter would be particularly useful for a developing a back pressure mechanism for physical actions.

@edwardalee Please feel free to edit this text for clarity so as to avoid creating a long conversational thread.

**Comment by Soroosh129** [comment]
I think instead of having `DEFER` or `DROP` you could have a second parameter to an action called `queue size`, which indicates how many future actions can be scheduled before you start dropping them.

**Comment by lsk567** [comment]
If I understand correctly, the proposed `MIT` is an attribute defined in the action declaration. So there are two questions about this: 1. Do we allow MIT to change during the program execution? 2. Can it be parameterized by a variable?

Also, an MIT violation by a logical action currently seems to be handled in the `schedule` function. What about the MIT violation by a physical action? Some `DEFER` or `DROP` policy should probably be there too. A potential place to put it is in the target properties, where the developer can decide based on the specific use case.

**Comment by lhstrh** [comment]
> I think instead of having `DEFER` or `DROP` you could have a second parameter to an action called `queue size`, which indicates how many future actions can be scheduled before you start dropping them. Not specifying it would be equivalent to the `DEFER` policy and setting it to 1 will be equivalent to the `DROP` policy.

We could, but I would consider what you describe fourth policy that does not have to come at the expense of the other three.

I think that it would be very useful to impose a bound on the size of the event queue, but I see this as something independent from the MIT violation policy. One could schedule an arbitrary `n` events in the future, all spaced by at least the specified MIT, meaning that no events would be dropped and the queue could grow boundlessly. Therefore, I think such constraint would have to be formulated independent from an MIT violation policy: something like "don't allow more than `x` events to be waiting in the queue per any action (or some action in particular)". Aside from what syntax we choose to specify such constraint, this raises another question: what should happen in `schedule` if the max capacity is reached?

**Comment by lhstrh** [comment]
> If I understand correctly, the proposed `MIT` is an attribute defined in the action declaration. So there are two questions about this: 1. Do we allow MIT to change during the program execution? 

No

> 2. Can it be parameterized by a variable?

Good question! I don't see a reason why not to allow parameters, except for the situation where parameters can be overridden at runtime, which I'm still not convinced is a good idea. The point is that we want to know the values of MITs at compile time for schedulability analysis.

> Also, an MIT violation by a logical action currently seems to be handled in the `schedule` function. What about the MIT violation by a physical action? The runtime system should probably execute some `DEFER` or `DROP` policy too. A potential place to put it is in the target properties, where the developer can decide based on the specific use case.

Physical actions and logical actions would be treated the same in terms of MIT enforcement. The only difference between logical and physical actions is the way the tag of the resulting event is computed. For logical actions `t_new = t + d`. For physical actions `t_new = max(t + d, T)`. Both are subject to the constraint that `t_new - t_last > m`, where `m` denotes the MIT. The specified policy determines what happens when this constraint is not satisfied. Note that `d` is currently composed of the intrinsic minimum delay of the action + the specified additional delay passed to `schedule`. In this proposal, `d` is simply the parameter passed to `schedule` since there no longer is a minimum delay (instead, there is the MIT parameter).

**Comment by edwardalee** [comment]
I updated the original proposal to say that, if given, the MIT must be strictly greater than zero.  An MIT of zero gets confusing and conflicts with the replacement policy when an event is rescheduled with the same tag.  Also, it's not clear what DEFER should do with an MIT of zero.

**Comment by lhstrh** [comment]
This has been implemented with very minimal syntax changes after all. See the `AsyncCallbackReplace.lf` and `AsyncCallbackDrop.lf` tests for reference. When a minimum interarrival time (MIT) has been specified, not only does this ensure a minimum spacing between events, it also enforces monotonicity of the tags of subsequent events. Now, non-monotonic behavior is only possible when no MIT is specified. This makes the implementation of policies `defer`, `drop`, and `replace` pretty straightforward. I will update the documentation in the wiki soon.
